src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1777 @@
+    void* des = tgt.node->from_end((to_src -1)->get_node_key().key_off - offset);
+    void* src = (to_src - 1)->get_node_val_ptr();
+    size_t len = from_src.get_index() == 0? (to_src -1)->get_node_key().key_off:
+                 (from_src-1)->get_node_val_ptr() - (to_src -1)->get_node_val_ptr();
+    memcpy(des, src, len);
+    memcpy(
+      tgt->get_node_key_ptr(), from_src->get_node_key_ptr(),
+      to_src->get_node_key_ptr() - from_src->get_node_key_ptr());
+    if ( from_src.get_index() == 0)
+      return;
+
+    omap_inner_key_t key = (from_src - 1)->get_node_key();
+    auto end_idx = tgt.get_index() + to_src.get_index() - from_src.get_index();
+    for (auto ite = tgt; ite.get_index() != end_idx; ite++) {
+       omap_inner_key_t node_key = ite->get_node_key();
+       node_key.key_off -= key.key_off;
+       ite->set_node_key(node_key);
+    }
+  }
+
+  /**
+   * append copy_from_foreign_ahead
+   *
+   * append another node head entries to this node back.
+   * [from_src, to_src) is another node entry range.
+   * tgt is this node entry to copy to.
+   * tgt and from_src must be from different nodes.
+   * from_src and to_src must be in the same node.
+   */
+  void append_copy_from_foreign_head(
+    iterator tgt,
+    const_iterator from_src,
+    const_iterator to_src) {
+    assert(tgt->node != from_src->node);
+    assert(to_src->node == from_src->node);
+    if (from_src == to_src)
+      return;
+
+    void* des = tgt.node->from_end((to_src -1)->get_node_key().key_off + (tgt - 1)->get_node_key().key_off);
+    void* src = (to_src - 1)->get_node_val_ptr();
+    size_t len = (to_src -1)->get_node_key().key_off;
+    memcpy(des, src, len);
+    memcpy(
+      tgt->get_node_key_ptr(), from_src->get_node_key_ptr(),
+      to_src->get_node_key_ptr() - from_src->get_node_key_ptr());
+    omap_inner_key_t key = (tgt - 1)->get_node_key();
+    auto end_idx = tgt.get_index() + to_src.get_index() - from_src.get_index();
+    for (auto ite = tgt; ite.get_index() != end_idx; ite++) {
+       omap_inner_key_t node_key = ite->get_node_key();
+       node_key.key_off += key.key_off;
+       ite->set_node_key(node_key);
+    }
+  }
+
+  /**
+   * local_move_back
+   *
+   * move this node entries range [from_src, to_src) back to tgt position.
+   *
+   * tgt, from_src, and to_src must be from the same node.
+   */
+  static void local_move_back(
+    omap_inner_key_t key,
+    iterator tgt,
+    iterator from_src,
+    iterator to_src) {
+    assert(tgt->node == from_src->node);
+    assert(to_src->node == from_src->node);
+    void* des = (to_src-1)->get_node_val_ptr() - key.key_len;
+    void* src = (to_src-1)->get_node_val_ptr();
+    size_t len = from_src.get_index() == 0?
+                 from_src->node->buf + BlockSize - (to_src-1)->get_node_val_ptr():
+                 (from_src-1)->get_node_val_ptr() - (to_src-1)->get_node_val_ptr();
+
+    memmove(des, src, len);
```

Comment:
This is only valid if the ranges do not overlap.  I think they may in some cases.  You probably mean memcpy?